### PR TITLE
net: lib: lwm2m_client_utils: Support for download URI

### PR DESCRIPTION
**Pull request by SeppoTakalo**
Download client inside nRF library already handles application
and modem firmwares, so it is logical choice to support the pkgURI
resource with it.

This bypasses the firmware_pull.c from Zephyr.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

### CODE REVIEW

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 1]
```diff
@@ -205,15 +215,103 @@ static int firmware_block_received_cb(uint16_t obj_inst_id,
...
+       }
+
+       if (fota_host) {
+               LOG_ERR("FOTA download already ongoing");
+               return -EBUSY;
+       }
+
+       fota_download_finnished = false;
+
+       char *s, *e;
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:278]: nitpicking: declaration and assignment could be done on a same line.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 2]
```diff
@@ -205,15 +215,103 @@ static int firmware_block_received_cb(uint16_t obj_inst_id,
...
+       int len = e - s;
+       /* For host, I need to allocate space, as I need to copy the substring */
+       fota_host = k_malloc(len + 1);
+       if (!fota_host) {
+               LOG_ERR("Failed to allocate memory");
+               return -ENOMEM;
+       }
+       strncpy(fota_host, s, len);
+       fota_host[len] = 0;
+       k_work_schedule(&download_work, REBOOT_DELAY);
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:301]: Any specific reason to use REBOOT_DELAY instead of saying K_SECONDS(1)?

**Review by VeijoPesonen — APPROVED** [review]
Looks good to me

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 3]
```diff
@@ -205,15 +215,111 @@ static int firmware_block_received_cb(uint16_t obj_inst_id,
...
+		break;
+	}
+	k_free(fota_host);
+	fota_host = NULL;
+	fota_path = NULL;
+}
+
+static void start_fota_download(struct k_work *work)
+{
+	int ret = fota_download_start(fota_host, fota_path, 16842753, NULL, 0);
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c]: The sec tag is now configurable.

**Review by rlubos — APPROVED** [review]
Thanks. We should add an entry in the release notes covering this change though.

**Comment by SeppoTakalo** [comment]
Added entry to release notes as well.

**Comment by rlubos** [comment]
@SeppoTakalo Rebase please.

File: doc/nrf/releases/release-notes-changelog.rst [hunk 4]
```diff
@@ -38,6 +38,10 @@ nRF9160
     * LwM2M sensor objects now uses the actual sensors available to the Thingy:91. If the nRF9160 DK is used, it uses simulated sensors instead.
     * Added possibility to poll sensors and notify the server if the measured changes are large enough.
 
+  * :ref:`lib_lwm2m_client_utils` library:
+
+    * Added support for Firmware update object to use :ref:`lib_fota_download` library for downloading firmware images.
```
Comment (VeijoPesonen) [inline doc/nrf/releases/release-notes-changelog.rst:43]: Should it be stated that this is about HTTP.

Comment (SeppoTakalo) [reply doc/nrf/releases/release-notes-changelog.rst:43]: It is not limited to any protocol, at least not on this side.
Download is handled by [Download client](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/networking/download_client.html#protocols).
> The library supports HTTP, HTTPS (TLS 1.2), CoAP, and CoAPS (DTLS 1.2) over IPv4 and IPv6.

So I think it is enough that we point to the library that handles it. That library can be extended in the future.

**Comment by umapraseeda** [comment]
I shall add a commit.

**Review by umapraseeda — APPROVED** [review]

**Comment by rlubos** [comment]
Please squash the documentation review commit. And please rebase along the occasion, that changelog file is driving me crazy (I guess not only me...).

**Comment by SeppoTakalo** [comment]
Rebased & Squashed.

**Pull request merged**